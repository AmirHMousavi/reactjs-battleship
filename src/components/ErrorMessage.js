import React from "react";
import { Message } from "semantic-ui-react";
import PropTypes from "prop-types";

/**
 * An Error component, shows one message [just to warn palyer]
 * TODO: Handle better with ability of specified message by sending error number or error message as props
 * @param {error} props
 * @see component:ErrorMessage
 */
const ErrorMessage = props => {
  const { errorMessage } = props;
  return (
    <Message negative>
      <Message.Header>{errorMessage.header}</Message.Header>
      {errorMessage.body}
    </Message>
  );
};

ErrorMessage.propTypes = {
  errorMessage: PropTypes.object.isRequired // eslint-disable-line
};
export default ErrorMessage;
