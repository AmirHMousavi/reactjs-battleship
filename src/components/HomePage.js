import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { Button, Container } from "semantic-ui-react";
import PropTypes from "prop-types";
import { createNewPlayer } from "../actions/AllActions";

export class HomePage extends React.Component {
  startGame = () => {
    const { createNewPlayer: cnp, history } = this.props;
    cnp();
    history.push("/GameInit");
  };

  render() {
    return (
      <div>
        <Container textAlign="center">
          <Button.Group vertical>
            <Button id="startgame" color="blue" onClick={this.startGame}>
              START GAME
            </Button>
            <br />
            <Button id="scoreboard" color="teal" as={Link} to="/ScoreBoard">
              SCORE BOARD
            </Button>
            <br />
            <Button id="gamerules" color="yellow" as={Link} to="/Rules">
              GAME RULES
            </Button>
          </Button.Group>
        </Container>
      </div>
    );
  }
}

HomePage.propTypes = {
  createNewPlayer: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default connect(
  null,
  { createNewPlayer }
)(HomePage);
