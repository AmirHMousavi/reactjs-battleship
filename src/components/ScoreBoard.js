import React, { Component } from "react";
import {
  Table,
  Container,
  Header,
  Image,
  Button,
  Divider
} from "semantic-ui-react";
import { Link } from "react-router-dom";
import avatar from "../assets/avatar.png";

class ScoreBoard extends Component {
  state = {};

  render() {
    const scores = JSON.parse(localStorage.getItem("BatltleShipScores"));
    return (
      <div>
        <Container textAlign="center">
          <Header> SCORE BOARD</Header>
          <Table color="blue" celled textAlign="center">
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell>Player</Table.HeaderCell>
                <Table.HeaderCell>Wins</Table.HeaderCell>
                <Table.HeaderCell>Loses</Table.HeaderCell>
                <Table.HeaderCell>TotalGames</Table.HeaderCell>
              </Table.Row>
            </Table.Header>

            <Table.Body>
              {Object.keys(scores).map(key => (
                <Table.Row key={key}>
                  <Table.Cell>
                    <Header as="h4" image>
                      <Image src={avatar} rounded size="mini" />
                      <Header.Content>{scores[key].name}</Header.Content>
                    </Header>
                  </Table.Cell>
                  <Table.Cell>{scores[key].wins}</Table.Cell>
                  <Table.Cell>{scores[key].loses}</Table.Cell>
                  <Table.Cell>
                    {scores[key].wins + scores[key].loses}
                  </Table.Cell>
                </Table.Row>
              ))}
            </Table.Body>
          </Table>
          <Divider />
          <Button primary as={Link} to="/">
            Go To HomePage
          </Button>
        </Container>
      </div>
    );
  }
}

export default ScoreBoard;
