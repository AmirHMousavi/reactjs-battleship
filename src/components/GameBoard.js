import React, { Component } from "react";
import { Table, Divider, Button } from "semantic-ui-react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import _ from "lodash";
import ErrorMessage from "./ErrorMessage";
import {
  setShipLocation,
  resetBoard,
  changeMyTurnToPlay,
  reduceTheHealthOfTheShip
} from "../actions/AllActions";

/**
 * @classdesc GameBoard can be in two modes;
 * 1.Init mode where players are placing their ships on the board.
 * 2.Battle mode where players are shooting on each others boards.
 * this class is rendered as child of @see GameInit and @see Battle
 * @todo The class has many functions, better to seperate them in an action file and import them here.
 */
class GameBoard extends Component {
  state = {
    columnsENUM: ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J"],
    rowsENUM: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    error: {},
    mode: "",
    targetShots: [],
    lostShots: []
  };

  componentWillMount() {
    const { mode: receivedMode } = this.props;
    const { mode } = this.state;
    if (mode !== receivedMode) {
      this.setState({ mode: receivedMode });
    }
  }

  /**
   * when the game is in initialization mode this method handles clicks on the board
   * @param {number} rowNum
   * @param {string} colName
   */
  handleCellClickAtGameInit = (rowNum, colName) => {
    // destruct all needed values for ease
    const { players, playerNumber, setShipLocation: setLocation } = this.props;
    const { selectedShip, orientation } = players[playerNumber];
    const { size, key } = selectedShip;
    const { columnsENUM, rowsENUM, mode } = this.state;
    // get the index number of the column letter/name
    const colNum = columnsENUM.indexOf(colName);

    if (_.isEqual(mode, "InitMode")) {
      if (this.canPlaceItHere(rowNum, colName)) {
        const temp = [];
        if (orientation) {
          for (let i = 0; i < size; i += 1) {
            temp.push(rowNum + columnsENUM[colNum + i]);
          }
          setLocation(playerNumber, key, temp);
        }
        if (!orientation) {
          for (let i = 0; i < size; i += 1) {
            temp.push(rowsENUM[rowNum + i] + colName);
          }
          setLocation(playerNumber, key, temp);
        }
      } else {
        /** a general message for both not selected ship or the cells are occupide
         @todo: shif it to @function canPlaceItHere() with sepcific message */
        this.setState({
          error: {
            header: "BoardError!",
            body: "Either Ship is Not selected or Borad Cell Is Occupied"
          }
        });
        this.clrearError();
      }
    }
  };

  /**
   * when the game is in battle mode this method handles clicks on the board
   * @param {number} rowNum
   * @param {string} colName
   */
  handleCellClickAtBattle = (rowNum, colName) => {
    const {
      players,
      playerNumber,
      enemyNumber,
      changeMyTurnToPlay: cmttp,
      reduceTheHealthOfTheShip: rthots
    } = this.props;
    const { myTurnToPlay, username } = players[playerNumber];

    if (myTurnToPlay) {
      this.setState({
        error: {
          header: `It is ${username}'s turn to playe`,
          body: `Click On Enemy Board :)`
        }
      });
      this.clrearError();
    } else {
      // target shot
      if (this.isOccupied(rowNum, colName)) {
        const { targetShots: ts } = this.state;
        ts.push(rowNum + colName);
        this.setState({ targetShots: ts });
        const { ships } = players[playerNumber];
        const shotShip = ships.filter(ship =>
          ship.location.includes(rowNum + colName)
        );
        rthots(playerNumber, shotShip[0].key);
      }
      // lost shot
      else {
        const { lostShots: ls } = this.state;
        ls.push(rowNum + colName);
        this.setState({ lostShots: ls });
      }
      cmttp(playerNumber);
      cmttp(enemyNumber);
    }
  };

  /** Resets  ships[] Array in players state to its initial value (where the locations are empty)
   * @fires resetBoard
   */
  resetBoard() {
    const { resetBoard: resb, playerNumber } = this.props;
    resb(playerNumber);
  }

  /** a timeout that set error state to empty after 4 seconds
   * @fires setState
   */
  clrearError() {
    setTimeout(() => {
      this.setState({ error: {} });
    }, 4000);
  }

  /**
   * checks if given cell (rowNum+colName) is occupied
   * @param {number} rowNum
   * @param {string} colName
   */
  isOccupied(rowNum, colName) {
    const { players, playerNumber } = this.props;
    const { ships } = players[playerNumber];

    return ships.some(key => key.location.includes(rowNum + colName));
  }

  /**
   * cehcks if a ship can be placed at the given cell
   * by checking distance to borders of the board
   * and if the cells needed are already occupied
   * @fires isOccupied
   * @param {number} rowNum
   * @param {string} colName
   */
  canPlaceItHere(rowNum, colName) {
    const { players, playerNumber } = this.props;
    const { orientation, selectedShip } = players[playerNumber];
    const { size } = selectedShip;
    const { columnsENUM, rowsENUM } = this.state;
    const colNumFromName = columnsENUM.indexOf(colName);

    if (_.isEmpty(selectedShip)) return false;
    if (orientation) {
      if (size > columnsENUM.length - colNumFromName) return false;
      for (let index = 0; index < size; index += 1) {
        if (this.isOccupied(rowNum, columnsENUM[colNumFromName + index]))
          return false;
      }
    }
    if (!orientation) {
      if (size > rowsENUM.length - rowNum) return false;
      for (let index = 0; index < size; index += 1) {
        if (this.isOccupied(rowsENUM[rowNum + index], colName)) return false;
      }
    }
    return true;
  }

  render() {
    const {
      columnsENUM,
      rowsENUM,
      error,
      mode,
      lostShots,
      targetShots
    } = this.state;
    return (
      <div>
        <div>
          {/* ----- Rendering a 10x10 Table---------- */}
          <Table celled definition unstackable fixed>
            <Table.Header>
              <Table.Row textAlign="center">
                <Table.HeaderCell />
                {columnsENUM.map(colName => (
                  <Table.HeaderCell key={colName}>{colName}</Table.HeaderCell>
                ))}
              </Table.Row>
            </Table.Header>
            {/** should call the onClick as a function
             the cell passes rowNumber and columnnumber
             @see https://github.com/Semantic-Org/Semantic-UI-React/pull/829#issuecomment-416953761 */}
            <Table.Body>
              {rowsENUM.map(rowNum => (
                <Table.Row key={rowNum} textAlign="center" color="blue">
                  <Table.Cell key={rowNum}>{rowNum}</Table.Cell>
                  {columnsENUM.map(colName => {
                    if (_.isEqual(mode, "InitMode"))
                      return (
                        <Table.Cell
                          key={rowNum + colName}
                          selectable
                          onClick={() =>
                            this.handleCellClickAtGameInit(rowNum, colName)
                          }
                          bgcolor={
                            this.isOccupied(rowNum, colName)
                              ? "#331a00"
                              : "#b3e6ff"
                          }
                        />
                      );
                    if (_.isEqual(mode, "BattleMode")) {
                      let color = "";
                      if (lostShots.includes(rowNum + colName))
                        color = "#b3e6ff";
                      else if (targetShots.includes(rowNum + colName))
                        color = "red";
                      return (
                        <Table.Cell
                          key={rowNum + colName}
                          onClick={() =>
                            this.handleCellClickAtBattle(rowNum, colName)
                          }
                          bgcolor={color}
                        >
                          {rowNum + colName}
                        </Table.Cell>
                      );
                    }
                    return <div />;
                  })}
                </Table.Row>
              ))}
            </Table.Body>
          </Table>
          {/*------------------------------------*/}
          <Divider />
          {_.isEqual(mode, "InitMode") ? (
            <Button color="yellow" onClick={() => this.resetBoard()}>
              Reset Board
            </Button>
          ) : (
            <div />
          )}
        </div>
        <div>
          {!_.isEmpty(error) ? <ErrorMessage errorMessage={error} /> : <div />}
        </div>
      </div>
    );
  }
}

/** Find Actions in src/actions/AllActions  */
GameBoard.propTypes = {
  setShipLocation: PropTypes.func.isRequired,
  mode: PropTypes.string.isRequired,
  resetBoard: PropTypes.func.isRequired,
  changeMyTurnToPlay: PropTypes.func.isRequired,
  reduceTheHealthOfTheShip: PropTypes.func.isRequired,
  // as players object is big, using 'shape' is inconvenient. disable ESLint here
  players: PropTypes.object.isRequired, // eslint-disable-line
  playerNumber: PropTypes.number.isRequired,
  enemyNumber: PropTypes.number.isRequired
};

const mapStateToProps = state => ({
  players: state.players
});

export default connect(
  mapStateToProps,
  { setShipLocation, resetBoard, changeMyTurnToPlay, reduceTheHealthOfTheShip }
)(GameBoard);
