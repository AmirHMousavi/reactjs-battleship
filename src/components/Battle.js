import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import _ from "lodash";
import { Grid, Header, Button, Icon, Message } from "semantic-ui-react";
import GameBoard from "./GameBoard";
import { changeMyTurnToPlay, cancelGame } from "../actions/AllActions";

class Battle extends Component {
  state = { gameOver: {} };

  componentWillMount() {
    const { players, changeMyTurnToPlay: cmttp } = this.props;
    // just to make sure it is no ones turn to play
    Object.keys(players).forEach(
      key => (players[key].myTurnToPlay ? cmttp(key) : {})
    );
    // now give player one right to play
    cmttp(1);
  }

  componentDidUpdate() {
    const { gameOver } = this.state;
    // player which all his ships hited, is the looser
    const looserPlayer = Number(this.getHitedShips()[0]);
    if (looserPlayer > 0) {
      let winerPlayer;
      if (looserPlayer === 1) {
        winerPlayer = 2;
      } else winerPlayer = 1;
      const go = { looser: looserPlayer, winner: winerPlayer };
      // check if gameOverState is changed from prevSate. avoid infinitLoop
      if (!_.isEqual(gameOver, go)) {
        this.setState({ gameOver: go }); // eslint-disable-line
      }
      this.addResultToscoreBoard();
    }
  }

  /**
   * checks which player has all ships with health of 0
   * @returns Object Key which indicates player number
   */
  getHitedShips = () => {
    const { players } = this.props;
    return Object.keys(players).filter(
      key => players[key].ships.filter(ship => ship.health <= 0).length >= 5
    );
  };

  /**
   * clear thestore state and redirect to HomePage
   */
  cancel = () => {
    const { history, cancelGame: cg } = this.props;
    cg();
    history.push("/");
  };

  addResultToscoreBoard() {
    const { players } = this.props;
    const { gameOver } = this.state;
    // during development first render was returing empty gameOver so lets put a condition here ;)
    if (!_.isEmpty(gameOver)) {
      const winerName = players[gameOver.winner].username;
      const looserName = players[gameOver.looser].username;
      const scores = JSON.parse(localStorage.getItem("BatltleShipScores"));
      if (Object.keys(scores).includes(winerName)) {
        scores[winerName].wins += 1;
      } else {
        scores[winerName] = { name: winerName, wins: 1, loses: 0 };
      }
      if (Object.keys(scores).includes(looserName)) {
        scores[looserName].loses += 1;
      } else {
        scores[looserName] = { name: looserName, wins: 0, loses: 1 };
      }
      localStorage.setItem("BatltleShipScores", JSON.stringify(scores));
    }
  }

  render() {
    const { players } = this.props;
    const { gameOver } = this.state;
    return (
      <div>
        <Grid columns={2} divided>
          <Grid.Column stretched>
            <Header as="h3" textAlign="center" color="teal">
              {players[1].username}
              &acute;s Board
            </Header>
            <GameBoard playerNumber={1} enemyNumber={2} mode="BattleMode" />
          </Grid.Column>
          <Grid.Column stretched>
            <Header as="h3" textAlign="center" color="teal">
              {players[2].username}
              &acute;s Board
            </Header>
            <GameBoard playerNumber={2} enemyNumber={1} mode="BattleMode" />
          </Grid.Column>
        </Grid>
        <br />
        {_.isEmpty(gameOver) ? (
          <Button
            color="red"
            icon
            labelPosition="right"
            type="button"
            onClick={() => this.cancel()}
          >
            Exit Game
            <Icon name="left arrow" />
          </Button>
        ) : (
          <br />
        )}
        {!_.isEmpty(gameOver) ? (
          <div>
            <Message color="blue">
              <Message.Header>GAME OVER!</Message.Header>
              {players[gameOver.looser].username} Lost All The Ships,{" "}
              {players[gameOver.winner].username} Is The Winner
            </Message>
            <Button
              color="blue"
              icon
              labelPosition="right"
              type="button"
              onClick={() => this.cancel()}
            >
              Start a New Game
              <Icon name="left arrow" />
            </Button>
          </div>
        ) : (
          <div />
        )}
      </div>
    );
  }
}

Battle.propTypes = {
  changeMyTurnToPlay: PropTypes.func.isRequired,
  cancelGame: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  players: PropTypes.object.isRequired // eslint-disable-line
};

const mapStateToProps = state => ({
  players: state.players
});
export default connect(
  mapStateToProps,
  { changeMyTurnToPlay, cancelGame }
)(Battle);
