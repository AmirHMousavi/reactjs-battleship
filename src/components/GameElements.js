import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import {
  Form,
  FormField,
  Dropdown,
  Button,
  Icon,
  Grid
} from "semantic-ui-react";
import _ from "lodash";
import {
  changeOrientation,
  nameInputOnChange,
  shipSelected,
  cancelGame
} from "../actions/AllActions";
import GameBoard from "./GameBoard";
import ErrorMessage from "./ErrorMessage";

export class GameElements extends Component {
  state = { errors: {} };

  onSubmit = () => {
    const { whatNext } = this.props;
    const errors = this.validate();
    if (!_.isEmpty(errors)) {
      /** @see https://github.com/jaredpalmer/formik/issues/772 */
      this.setState({ errors });
      setTimeout(() => {
        this.setState({ errors: {} });
      }, 4000);
    }
    if (Object.keys(errors).length === 0) {
      whatNext();
    }
  };

  validate = () => {
    const { players, playerNumber } = this.props;
    const { username, ships } = players[playerNumber];
    const errors = {};
    if (_.isEmpty(username)) {
      errors.header = "FromError!";
      errors.body = "username can't be blank";
    }
    ships.forEach(ship => {
      if (_.isEmpty(ship.location)) {
        errors.header = "FromError!";
        errors.body = "All ships must be located";
      }
    });
    return errors;
  };

  render() {
    const {
      players,
      playerNumber,
      changeOrientation: co,
      nameInputOnChange: nameInputOnChangeAlt,
      shipSelected: shipSelectedAlt,
      cancel: cg
    } = this.props;
    const currentPlayer = players[playerNumber];
    const { username, ships, orientation } = currentPlayer;
    const { errors } = this.state;

    return (
      <Grid>
        <Grid.Column width={4}>
          <Form id="from" onSubmit={this.onSubmit}>
            {/** @todo check if username already exists  */}
            <Form.Field>
              <label htmlFor="username">
                username
                <input
                  id="username"
                  type="text"
                  placeholder="e.g. John"
                  value={username}
                  onChange={e =>
                    nameInputOnChangeAlt(playerNumber, e.target.value)
                  }
                />
              </label>
            </Form.Field>
            <FormField>
              <Button
                id="orientation"
                primary
                icon
                type="button"
                onClick={() => co(playerNumber)}
              >
                <Icon name="retweet" />
                &nbsp;
                {orientation ? "Horizental" : "Vertical"}
              </Button>
            </FormField>
            <FormField>
              <Dropdown
                id="selectedship"
                placeholder="Select Ship"
                options={ships}
                onChange={(e, { value }) =>
                  shipSelectedAlt(playerNumber, ships[value])
                }
                selection
                item
              />
            </FormField>
            <FormField>
              <Form.Group>
                <Button primary type="submit" icon labelPosition="right">
                  Save &amp; Next
                  <Icon name="right arrow" />
                </Button>
              </Form.Group>
            </FormField>

            <FormField>
              <Button
                color="red"
                icon
                labelPosition="right"
                type="button"
                onClick={() => cg()}
              >
                Cancel
                <Icon name="left arrow" />
              </Button>
            </FormField>
          </Form>
        </Grid.Column>
        <Grid.Column width={12}>
          {/* Here we call the board , game board has access to the store */}
          <GameBoard
            playerNumber={playerNumber}
            enemyNumber={-1}
            mode="InitMode"
          />
        </Grid.Column>
        {!_.isEmpty(errors) ? <ErrorMessage errorMessage={errors} /> : <div />}
      </Grid>
    );
  }
}

/** Find Actions in src/actions/PlayerActions  */
GameElements.propTypes = {
  changeOrientation: PropTypes.func.isRequired,
  nameInputOnChange: PropTypes.func.isRequired,
  shipSelected: PropTypes.func.isRequired,
  cancel: PropTypes.func.isRequired,
  whatNext: PropTypes.func.isRequired,
  players: PropTypes.object.isRequired, // eslint-disable-line
  playerNumber: PropTypes.number.isRequired
};

const mapStateToProps = state => ({
  players: state.players
});

export default connect(
  mapStateToProps,
  {
    changeOrientation,
    nameInputOnChange,
    shipSelected,
    cancelGame
  }
)(GameElements);
