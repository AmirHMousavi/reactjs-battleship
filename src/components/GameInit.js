import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { Divider } from "semantic-ui-react";
import _ from "lodash";
import GameElements from "./GameElements";
import { createNewPlayer, cancelGame } from "../actions/AllActions";

class GameInit extends Component {
  whatNext = () => {
    const {
      players,
      createNewPlayer: createNewPlayerAlt,
      history
    } = this.props;
    const playerNumber = Object.keys(players).length;
    // if we already have 2 players, go to battle , else create new player
    if (playerNumber >= 2) {
      history.push("/Battle");
    } else {
      createNewPlayerAlt();
    }
  };

  cancel = () => {
    const { history, cancelGame: cg } = this.props;
    cg();
    history.push("/");
  };

  render() {
    const { players } = this.props;
    const playerNumber = Object.keys(players).length;
    if (!_.isEmpty(players))
      return (
        <div>
          <h2> Player {playerNumber} </h2>
          <Divider />
          <GameElements
            playerNumber={playerNumber}
            whatNext={this.whatNext}
            cancel={this.cancel}
          />
        </div>
      );
    return <div />;
  }
}

GameInit.propTypes = {
  createNewPlayer: PropTypes.func.isRequired,
  cancelGame: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  players: PropTypes.object.isRequired // eslint-disable-line
};

const mapStateToProps = state => ({
  players: state.players
});

export default connect(
  mapStateToProps,
  { createNewPlayer, cancelGame }
)(GameInit);
