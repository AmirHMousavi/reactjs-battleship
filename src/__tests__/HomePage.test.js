/* eslint-disable */
import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { HomePage } from "../components/HomePage";

let createNewPlayer, history, wrapper;

beforeEach(() => {
  createNewPlayer = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <HomePage createNewPlayer={createNewPlayer} history={history} />
  );
});

test("should render the HomePage Component", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
  wrapper.find("#startgame").prop("onClick")();
  expect(history.push).toHaveBeenLastCalledWith("/GameInit");
});
