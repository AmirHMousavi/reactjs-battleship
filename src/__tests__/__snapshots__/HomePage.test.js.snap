// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the HomePage Component 1`] = `
<div>
  <Container
    textAlign="center"
  >
    <ButtonGroup
      vertical={true}
    >
      <Button
        as="button"
        color="blue"
        id="startgame"
        onClick={[Function]}
        role="button"
      >
        START GAME
      </Button>
      <br />
      <Button
        as={[Function]}
        color="teal"
        id="scoreboard"
        role="button"
        to="/ScoreBoard"
      >
        SCORE BOARD
      </Button>
      <br />
      <Button
        as={[Function]}
        color="yellow"
        id="gamerules"
        role="button"
        to="/Rules"
      >
        GAME RULES
      </Button>
    </ButtonGroup>
  </Container>
</div>
`;
